pkg/isb/serde.go

```diff
@@ -1,431 +1,126 @@
-/*
-Copyright 2022 The Numaproj Authors.
-
-Licensed under the Apache License, Version 2.0 (the "License");
-you may not use this file except in compliance with the License.
-You may obtain a copy of the License at
-
-    http://www.apache.org/licenses/LICENSE-2.0
-
-Unless required by applicable law or agreed to in writing, software
-distributed under the License is distributed on an "AS IS" BASIS,
-WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-See the License for the specific language governing permissions and
-limitations under the License.
-*/
-
 package isb
 
 import (
-	"bytes"
-	"encoding/binary"
-	"fmt"
-	"time"
-)
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/timestamppb"
 
-type messageInfoPreamble struct {
-	EventEpoch int64
-	IsLate     bool
-}
+	"github.com/numaproj/numaflow/pkg/apis/proto/isb"
+)
 
-// MarshalBinary encodes MessageInfo to the binary format
-func (p MessageInfo) MarshalBinary() (data []byte, err error) {
-	var buf = new(bytes.Buffer)
-	var preamble = messageInfoPreamble{
-		EventEpoch: p.EventTime.UnixMilli(),
-		IsLate:     p.IsLate,
-	}
-	if err = binary.Write(buf, binary.LittleEndian, preamble); err != nil {
-		return nil, err
-	}
-	return buf.Bytes(), nil
-}
+// MarshalBinary encodes Message to proto bytes.
+func (m Message) MarshalBinary() ([]byte, error) {
+	pb := &isb.Message{
+		Header: &isb.Header{
+			MessageInfo: &isb.MessageInfo{
+				EventTime: timestamppb.New(m.Header.MessageInfo.EventTime),
+				IsLate:    m.Header.MessageInfo.IsLate,
+			},
+			Kind: isb.MessageKind(m.Header.Kind),
+			Id: &isb.MessageID{
+				VertexName: m.Header.ID.VertexName,
+				Offset:     m.Header.ID.Offset,
+				Index:      m.Header.ID.Index,
+			},
+			Keys:    m.Header.Keys,
+			Headers: m.Header.Headers,
+		},
+		Body: &isb.Body{
+			Payload: m.Body.Payload,
+		},
+	}
+	return proto.Marshal(pb)
+}
+
+// UnmarshalBinary decodes Message from the proto bytes.
+// Even though the header and body are pointers in the proto, we don't need
+// nil checks because while marshalling, we always set the default values.
+func (m *Message) UnmarshalBinary(data []byte) error {
+	pb := &isb.Message{}
+	if err := proto.Unmarshal(data, pb); err != nil {
+		return err
+	}
+
+	m.Header = Header{
```

Comment:
```suggestion
	// no nil checks are performed because Marshalling and Unmarshalling
	// are completely controlled by numaflow. It is best not to do nil checks
	// and get panics so we can fix the root cause.
	
	m.Header = Header{
```